### PR TITLE
Added ability to convert from datetime.date to system_clock::time_point

### PR DESCRIPTION
**Pull request by phil-zxx**
Added ability to convert from Python datetime.date to C++ system_clock::time_point

As raised here: https://github.com/pybind/pybind11/issues/1844

### CODE REVIEW

File: include/pybind11/chrono.h [hunk 1]
```diff
@@ -119,6 +119,19 @@ template <typename Duration> class type_caster<std::chrono::time_point<std::chro
...
+            cal.tm_min   = 0;
+            cal.tm_hour  = 0;
+            cal.tm_mday  = PyDateTime_GET_DAY(src.ptr());
+            cal.tm_mon   = PyDateTime_GET_MONTH(src.ptr()) - 1;
+            cal.tm_year  = PyDateTime_GET_YEAR(src.ptr()) - 1900;
+            cal.tm_isdst = -1;
+
+            value = system_clock::from_time_t(std::mktime(&cal));
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (phil-zxx) [reply include/pybind11/chrono.h]: Would `PyDateTime_TIME_GET_HOUR`, etc. even work? Given the input is a `datetime.date` (and thus does not have any time information).

File: include/pybind11/chrono.h [hunk 2]
```diff
@@ -119,6 +119,19 @@ template <typename Duration> class type_caster<std::chrono::time_point<std::chro
             value = system_clock::from_time_t(std::mktime(&cal)) + microseconds(PyDateTime_DATE_GET_MICROSECOND(src.ptr()));
             return true;
         }
+        else if (PyDate_Check(src.ptr())) {
```
[the remark this answers is not in the document]
Comment (wjakob) [reply include/pybind11/chrono.h]: yep.

File: include/pybind11/chrono.h [hunk 3]
```diff
@@ -118,6 +118,30 @@ template <typename Duration> class type_caster<std::chrono::time_point<std::chro
 
             value = system_clock::from_time_t(std::mktime(&cal)) + microseconds(PyDateTime_DATE_GET_MICROSECOND(src.ptr()));
             return true;
+        } else if (PyDate_Check(src.ptr())) {
+            std::tm cal;
```
Comment (wjakob) [inline include/pybind11/chrono.h]: Could you move the std::tm object to the parent scope?

File: include/pybind11/chrono.h [hunk 4]
```diff
@@ -118,6 +118,30 @@ template <typename Duration> class type_caster<std::chrono::time_point<std::chro
...
+            cal.tm_sec   = PyDateTime_TIME_GET_SECOND(src.ptr());
+            cal.tm_min   = PyDateTime_TIME_GET_MINUTE(src.ptr());
+            cal.tm_hour  = PyDateTime_TIME_GET_HOUR(src.ptr());
+            cal.tm_mday  = 1;   // This date (day, month, year) = (1, 0, 70)
+            cal.tm_mon   = 0;   // represents 1-Jan-1970, which is the first
+            cal.tm_year  = 70;  // earliest available date for Python's datetime
+            cal.tm_isdst = -1;
+
+            value = system_clock::from_time_t(std::mktime(&cal)) + microseconds(PyDateTime_TIME_GET_MICROSECOND(src.ptr()));
```
Comment (wjakob) [inline include/pybind11/chrono.h]: Could you move this to the parent scope? (along with the other 2 cases where `value` is assigned.)

I suggest that all 3 cases write a `microsecond` variable that is simply zero in the `date`-only case.

Comment (phil-zxx) [reply include/pybind11/chrono.h]: There is `PyDateTime_DATE_GET_MICROSECOND` and `PyDateTime_TIME_GET_MICROSECOND`, so I think we have to keep this within each `if` block?

File: include/pybind11/chrono.h (hunk 3 above)
Comment (phil-zxx) [reply include/pybind11/chrono.h]: Move all `std::tm cal;` to the line before `if (PyDateTime_Check(src.ptr())) {`?

File: include/pybind11/chrono.h (hunk 4 above)
Comment (wjakob) [reply include/pybind11/chrono.h]: yep, but these can assign to a variable that is kept in the outer scope. The `from_time_t` call is then shared by all 3 branches.

**Pull request merged**